Make `airflow dags test` be able to execute Mapped Tasks

ashb opened the pull request:
In order to do this there were a few required:

- The BackfillJob needs to know about mapped tasks, both to expand them, and in order to update it's TI tracking
- The DebugExecutor, task_command.py, LocalTaskJob needed to "unmap" the mapped task to get the real operator back at the right time.

I was testing this with the following dag:

```python
from airflow import DAG
from airflow.decorators import task
from airflow.operators.python import PythonOperator
import pendulum

@task
def make_list():
    return list(map(lambda a: f'echo "{a!r}"', [1, 2, {'a': 'b'}]))

def consumer(*args):
     print(repr(args))

with DAG(dag_id='maptest', start_date=pendulum.DateTime(2022, 1, 18)) as dag:
    PythonOperator(task_id='consumer', python_callable=consumer).map(op_args=make_list())
```

It can't "unmap" decorated operators successfully yet, so we're using old-school PythonOperator

We also just pass the whole value to the operator, not just the current mapping value(s)


<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

ashb committed:
Make `airflow dags test` be able to execute Mapped Tasks

In order to do this there were two steps required:

- The BackfillJob needs to know about mapped tasks, both to expand them,
  and in order to update it's TI tracking
- The DebugExecutor needed to "unmap" the mapped task to get the real
  operator bacj

I was testing this with the following dag:

```
from airflow import DAG
from airflow.decorators import task
from airflow.operators.python import PythonOperator
import pendulum

@task
def make_list():
    return list(map(lambda a: f'echo "{a!r}"', [1, 2, {'a': 'b'}]))

def consumer(*args):
     print(repr(args))

with DAG(dag_id='maptest', start_date=pendulum.DateTime(2022, 1, 18)) as dag:
    PythonOperator(task_id='consumer', python_callable=consumer).map(op_args=make_list())
```

It can't "unmap" decorated operators successfully yet, so we're using
old-school PythonOperator

We also just pass the whole value to the operator, not just the current
mapping value(s)

uranusjr commented on airflow/executors/debug_executor.py:
```diff
@@ -76,6 +79,8 @@ def _run_task(self, ti: TaskInstance) -> bool:
         try:
             params = self.tasks_params.pop(ti.key, {})
+            if ti.task.is_mapped:
+                ti.task = cast("MappedOperator", ti.task).unmap()
```
I wonder if it’d be nicer to also implement `unmap()` on `BaseOperator` and have it simply `return self`.

ashb replied:
I made it raise an exception (as it should never be called?) but exist for typing.

I could have it `return self` and then remove the `if task.is_mapped` -- WDYT?

uranusjr replied:
My OOP training tells me polymorphism is better than an if clause, so my vote goes to `return self`. But ultimately I guess either makes sense (depending on how your mental model is of the logic).

uranusjr commented on airflow/jobs/backfill_job.py:
```diff
@@ -623,15 +650,25 @@ def tabulate_ti_keys_set(set_ti_keys: Set[TaskInstanceKey]) -> str:
+            tis_values = (
+                (ti.dag_id, ti.task_id, ti.run_id, ti.map_index, ti.try_number) for ti in sorted_tis
+            )
+            return tabulate(tis_values, headers=["DAG ID", "Task ID", "Run ID", "Map Index", "Try number"])
```
Probably should be discussed in more detail in the UX phase but I wonder if we should hide the map index column entirely if no tis are mapped, and display an empty cell (instead of -1) for non-mapped tis.

ashb replied:
Yeah, that would definitely be nicer

ashb replied:
Done.

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -1649,6 +1616,7 @@ def _trigger_rule_from_kwargs(self) -> int:
     @classmethod
     def from_operator(cls, operator: BaseOperator, mapped_kwargs: Dict[str, Any]) -> "MappedOperator":
         dag: Optional["DAG"] = getattr(operator, '_dag', None)
+        tg: Optional["TaskGroup"] = getattr(operator, 'task_group', None)
```
When are those attributes not available? We can probably make both of these always present on the class (defaulting to None) instead.

ashb replied:
Good shout.

ashb committed:
Always have a `task_group` property on DAGNodes

And since TaskGroup is a DAGNode, we don't need to store parent group
directly anymore -- it'll already be stored

ashb committed:
Add "integation" tests for running mapped tasks via BackfillJob

ashb committed:
Fix tests

uranusjr commented on airflow/serialization/serialized_objects.py:
```diff
@@ -1086,8 +1087,15 @@ def deserialize_task_group(
+
+        def set_ref(task: BaseOperator) -> BaseOperator:
+            task.task_group = ref
+            return task
```
```suggestion
        def set_ref(task: BaseOperator) -> BaseOperator:
            task.task_group = weakref.proxy(group)
            return task
```

ashb replied:
Is there a problem re-using the proxy?

uranusjr replied:
Not really; the two are actually equivalent because proxy is implemented to return the same thing from one object.

```pycon
>>> import weakref
>>> class A: pass
...
>>> a = A()
>>> id(weakref.proxy(a))
139967946164800
>>> id(weakref.proxy(a))
139967946164800
```

uranusjr commented on airflow/jobs/backfill_job.py:
```diff
@@ -266,6 +275,11 @@ def _manage_executor_state(self, running):
+                continue
+            if ti.state not in self.STATES_COUNT_AS_RUNNING:
+                for node in ti.task.mapped_dependants():
+                    assert isinstance(node, MappedOperator)
```
Is this guaranteed (and why?) or just a temporary measure to get pass Mypy?

ashb replied:
Temporary for now until we support mapping task groups (the DAG api supports it, but the runtime doesn't yet)

ashb replied:
The why: Because `expand_mapped_task` doesn't exist on BaseOperator, only MappedOperator

uranusjr replied:
Makes sense

ashb committed:
Update airflow/serialization/serialized_objects.py

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

ashb committed:
Only show "Map Index" in Backfill report when relevant

ashb committed:
If you want to test a DAG, it helps if you commit it

uranusjr commented on tests/models/__init__.py:
```diff
@@ -16,9 +16,9 @@
 
 DEFAULT_DATE = timezone.datetime(2016, 1, 1)
-TEST_DAGS_FOLDER = os.path.normpath(os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', 'dags'))
+TEST_DAGS_FOLDER = pathlib.Path(__file__).parent.parent / 'dags'
```
```suggestion
TEST_DAGS_FOLDER = pathlib.Path(__file__).parent.with_name('dags')
```

Nit

ashb replied:
Why's that?

ashb replied:
Oh I see, yeah that's nicer.

ashb committed:
Update tests/models/__init__.py

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>
